custom_components/versatile_thermostat/keep_alive.py

```diff
@@ -0,0 +1,43 @@
+"""Building blocks for the heater switch keep-alive feature.
+
+The heater switch keep-alive feature consists of regularly refreshing the state
+of directly controlled switches at a configurable interval (regularly turning the
+switch 'on' or 'off' again even if it is already turned 'on' or 'off'), just like
+the keep_alive setting of Home Assistant's Generic Thermostat integration:
+ https://www.home-assistant.io/integrations/generic_thermostat/
+"""
+
+from collections.abc import Awaitable, Callable
+from datetime import timedelta, datetime
+
+from homeassistant.core import HomeAssistant, CALLBACK_TYPE
+from homeassistant.helpers.event import async_track_time_interval
+
+
+class IntervalCaller:
+    """Repeatedly call a given async action function at a given regular interval.
+
+    Convenience wrapper around Home Assistant's `async_track_time_interval` function.
+    """
+
+    def __init__(self, hass: HomeAssistant, interval_sec: int) -> None:
+        self._hass = hass
+        self._interval_sec = interval_sec
+        self._remove_handle: CALLBACK_TYPE | None = None
+
+    def cancel(self):
+        if self._remove_handle:
+            self._remove_handle()
+            self._remove_handle = None
+
+    def set_async_action(self, action: Callable[[], Awaitable[None]]):
+        if not self._interval_sec:
+            return
+        self.cancel()
```

Comment:
Yes ... but ... for the future robustness of the code (many people as me will change this in the future) is it, in my opinion better to cancel() even if you think there is nothing to cancel.

This kind of programing "save my life" many time as I handles also non nominal cases. There is many piece of code like this (stop heater even if "normally" it is already turned off -> this is safety). Testing if xx <= 0 instead of if xxx == 0, ... all this patterns are for safety and robustness.

In that case, in the cancel method, if really there is nothing to cancel, then you do nothing. So, the cancel() method have to be think as "cancel_if_any", if you prefer.

Reply:
> I will do it if you insist further!

You haven’t yet insisted further, 🫣 so I am leaving it as it is. I understand that you are not quite happy with it, but I think I can live with that! 😄😇 The code is good.